### ISSUE TITLE
error modeling knn takes up all cores

### ISSUE THREAD

**Issue by dtm2117** [hms-dbmi/scde#51]
Hello,

I am trying to run the error modeling step on a sample with thousands of cells. I've increased K and min.nonfailed. But I find that even if I set n.cores = 1 , I have threads running on every core of the cluster. Furthermore I've never had the error modeling step finish. 

Any thoughts on this issue?

**Comment by pkharchenko** [comment]
- The number of cores might have to do with the environment, but just in case, could you please provide the exact command you’re using.
- In terms of runtime, it depends on the number of cells (linearly) and the number of genes. Filtering out low-expressed genes from the matrix should speed up calculations considerably.

**Comment by dtm2117** [comment]
Here is the command:
knn <- knn.error.models(cd_new_nodup, k = ncol(cd)/2, n.cores = 12, min.count.threshold = 1, min.nonfailed = 20, max.model.plots = 10)

dim of the matrix is ~ 13k genes but 4k cells. I've filtered out any genes that have no expression.

**Comment by dtm2117** [comment]
This is UMI data also

**Comment by pkharchenko** [comment]
For the runtime issue, I think k needs to be lowered considerably. To something like 50 or a 100. It just needs sufficient number of neighboring cells to calculate the few parameters for the error model. 2k cells would be definitely an overkill for that.
Not sure about the number of cores though. 
Can you see if doing something like 
pagoda:::papply( 1:1e2, function(x) rnorm(1e3), n.cores=12) 
also uses too many cores?
Best,
-peter.

**Comment by dtm2117** [comment]
Ok, I will try this. On the scde help page it says that k may need to be increased for 1000s of cells, which is why I kept the denominator low. 

Can't run that command because the pagoda library is not installed apparently. I thought it was installed along with SCDE package?

**Comment by pkharchenko** [comment]
Yes, I meant scde package. You'd need to increase the rnorm argument to take some sizeable amount of time.

-peter.

**Comment by dtm2117** [comment]
after increasing rnorm, it seems to be running on 12 cores only.

**Comment by dtm2117** [comment]
While the scde:::papply( 1:1e2, function(x) rnorm(1e3), n.cores=12) runs on only 12 cores, the knn parameter still uses all cores. 

Any ideas on why the discrepancy?
